plugins/modules/git_config.py

```diff
@@ -197,94 +217,127 @@ def main():
+    name = params['name'] or ''
+    unset = params['state'] == 'absent'
+    new_value = params['value'] or ''
+    add_mode = params['add_mode']
 
-    if params['scope']:
-        scope = params['scope']
-    elif params['list_all']:
-        scope = None
-    else:
-        scope = 'system'
+    scope = determine_scope(params)
+    cwd = determine_cwd(scope, params)
 
-    if params['state'] == 'absent':
-        unset = 'unset'
-        params['value'] = None
-    else:
-        unset = None
-
-    if params['value']:
-        new_value = params['value']
-    else:
-        new_value = None
+    base_args = [git_path, "config", "--includes"]
 
-    args = [git_path, "config", "--includes"]
-    if params['list_all']:
-        args.append('-l')
     if scope == 'file':
-        args.append('-f')
-        args.append(params['file'])
+        base_args.append('-f')
+        base_args.append(params['file'])
     elif scope:
-        args.append("--" + scope)
+        base_args.append("--" + scope)
+
+    list_args = list(base_args)
+
+    if params['list_all']:
+        list_args.append('-l')
+
     if name:
-        args.append(name)
+        list_args.append("--get-all")
+        list_args.append(name)
 
-    if scope == 'local':
-        dir = params['repo']
-    elif params['list_all'] and params['repo']:
-        # Include local settings from a specific repo when listing all available settings
-        dir = params['repo']
-    else:
-        # Run from root directory to avoid accidentally picking up any local config settings
-        dir = "/"
+    (rc, out, err) = module.run_command(list_args, cwd=cwd, expand_user_and_vars=False)
 
-    (rc, out, err) = module.run_command(args, cwd=dir, expand_user_and_vars=False)
     if params['list_all'] and scope and rc == 128 and 'unable to read config file' in err:
         # This just means nothing has been set at the given scope
         module.exit_json(changed=False, msg='', config_values={})
     elif rc >= 2:
         # If the return code is 1, it just means the option hasn't been set yet, which is fine.
-        module.fail_json(rc=rc, msg=err, cmd=' '.join(args))
+        module.fail_json(rc=rc, msg=err, cmd=' '.join(list_args))
+
+    old_values = out.rstrip().splitlines()
 
     if params['list_all']:
-        values = out.rstrip().splitlines()
         config_values = {}
-        for value in values:
+        for value in old_values:
             k, v = value.split('=', 1)
-            config_values[k] = v
+            if k in config_values:
+                config_values[k].append(v)
+            else:
+                config_values[k] = [v]
+
+        # All values in the dict are lists. If a list only contains one element, replace the
+        # list by the element itself.
+        for k, v in config_values.items():
+            if len(v) == 1:
+                config_values[k] = str(v[0])
         module.exit_json(changed=False, msg='', config_values=config_values)
```

Comment:
This is a breaking change for the format of the return value.

I think it's better to not modify the return values, but instead create a new module `git_config_info` that does querying. Then we can later on deprecate the `list_all` feature of this module (which should have never been added, see https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html#scoping-your-module-s).

Reply:
I tried to make this backward compatible in the sense of: 
* if the config only contains 1 key of the same name, nothing changes
* If the config contains more than 1 key of the same name, previousely, the module failed (because git failed), now a list of values is returned.

Would you see this as breaking change?

Reply:
Yes, because so far if multiple values were present, simply the last one of them was returned, instead of a list.

Reply:
@felixfontein what do you think about adding two new modules `git_config_get` and `git_config_set` which are doing the reading and writing and support multi-values as well. Then we could deprecate the `git_config` module untouched and remove it in the future.

Reply:
Or `git_config_write` and `git_config_read`

Reply:
I've pushed a new module `git_config_info` with a separate PR https://github.com/ansible-collections/community.general/pull/7587

Reply:
I reverted my change back to not break compatibility